src/main/kotlin/com/fasterxml/jackson/module/kotlin/KotlinValueInstantiator.kt

```diff
@@ -37,7 +41,12 @@ internal class KotlinValueInstantiator(src: StdValueInstantiator, private val ca
                 return@forEachIndexed
             }
 
-            val paramVal = buffer.getParameter(jsonProp)
+            val paramVal = if (buffer.hasParameter(jsonProp) || paramDef.isPrimitive()) {
```

Comment:
@zdevwu, could you describe the class you are trying to deserialize and the corresponding JSON, please? I'll dig into it.

Reply:
try just "{}" with the following class
class Type(
            val value: Optional<Int>
)

Reply:
There are a couple things that came to my mind during and after a debugging session:

GuavaModule works on assumption that all variables in Java can be set to null. When you mark a property as required using the JsonProperty annotation (as below) it fails.
```@JsonProperty(value = "value", required = true) Optional<Integer> value;```,
Jackson, even with GuavaModule installed, will fail to deserialize it from `{}` with an exception:
```
com.fasterxml.jackson.databind.exc.MismatchedInputException: Missing required creator property 'value' (index 0)
 at [Source: (String)"{}"; line: 1, column: 2]
```
What Kotlin module does is that it makes all variables required unless a default value is specified or the variable is marked nullable. It overrides the JsonProperty's required field. Which, I believe, makes sense for Kotlin. So, when you define a ```data class(val value: Optional<Int>)```, then the property `value` becomes required.
 So it seems to me that the modules try to achieve contradictory things.

 Even if the isPrimitive check was not there it would not work. Unfortunately, it is not that simple. Take a look at getParameter's code:
 https://github.com/FasterXML/jackson-databind/blob/master/src/main/java/com/fasterxml/jackson/databind/deser/impl/PropertyValueBuffer.java#L124
 Since there is no value passed in the JSON, the _findMissing function will be called. Inside, there is a check whether the property is required or not. In Kotlin it is, and in Java, it is not.
 
 To solve your problem you can:
 * rewrite your class to Java, where Kotlin module will not touch it
 * stop using Optional as property type and use it only on methods' return type. Jackson deserializes to either a value or a null and then you return Optional.of(value)
 * stop using Optional as it is not necessary for Kotlin (see https://kotlinlang.org/docs/reference/null-safety.html)

I am going to spend a bit more time on this to see if it can be fixed in a more elegant way. Maybe there is a way to check whether there is a custom deserializer available? Maybe GuavaModule should override the `required` flag?